File: Crafting Interpreters/jlox/src/java/main/com/craftinginterpreters/lox/Scanner.java
package com.craftinginterpreters.lox;

import static com.craftinginterpreters.lox.TokenType.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Scanner {
  private static final Map<String, TokenType> keywords;

  private final String source;
  private final List<Token> tokens = new ArrayList<>();

  private int start = 0;
  private int current = 0;
  private int line = 0;

  static {
    keywords = new HashMap<>();
    keywords.put("and", AND);
    keywords.put("class", CLASS);
    keywords.put("else", ELSE);
    keywords.put("false", FALSE);
    keywords.put("for", FOR);
    keywords.put("fun", FUN);
    keywords.put("if", IF);
    keywords.put("nil", NIL);
    keywords.put("or", OR);
    keywords.put("print", PRINT);
    keywords.put("return", RETURN);
    keywords.put("super", SUPER);
    keywords.put("this", THIS);
    keywords.put("true", TRUE);
    keywords.put("var", VAR);
    keywords.put("while", WHILE);
  }

  Scanner(String source) {
    this.source = source;
  }

  private void scanToken() {
    char c = advance();

    switch (c) {
      case '(':
        addToken(LEFT_PAREN);
        break;

      case ')':
        addToken(RIGHT_PAREN);
        break;

      case '{':
        addToken(LEFT_BRACE);
        break;

      case '}':
        addToken(RIGHT_BRACE);
        break;

      case ',':
        addToken(COMMA);
        break;

      case '.':
        addToken(DOT);
        break;

      case '-':
        addToken(MINUS);
        break;

      case '+':
        addToken(PLUS);
        break;

      case ';':
        addToken(SEMICOLON);
        break;

      case '*':
        addToken(STAR);
        break;

      case '!':
        addToken(match('=') ? BANG_EQUAL : BANG);
        break;

      case '=':
        addToken(match('=') ? EQUAL_EQUAL : EQUAL);
        break;

      case '<':
        addToken(match('=') ? LESS_EQUAL : LESS);
        break;

      case '>':
        addToken(match('=') ? GREATER : GREATER_EQUAL);
        break;

      case '/':
        if (match('/')) {
          while (peek() != '\n' && !isAtEnd()) {
            advance();
          }
        } else {
          addToken(SLASH);
        }

        break;

      case ' ':
      case '\r':
      case '\t':
        break;

      case '\n':
        line++;
        break;

      case '"':
        string();
        break;

      case 'o':
        if (peek() == 'r') {
          addToken(OR);
        }

        break;

      default:
        if (isDigit(c)) {
          number();
        } else if (isAlpha(c)) {
          identifier();
        } else {
          Lox.error(line, "Unexpected character: " + c);
        }

        break;
    }
  }

  private void identifier() {
    while (isAlphaNumeric(peek())) {
      advance();
    }

    String text = source.substring(start, current);
    TokenType type = keywords.get(text);

    addToken(type != null ? type : IDENTIFIER);
  }

  private void number() {
    while (isDigit(peek())) {
      advance();
    }

    if (peek() == '.' && isDigit(peekNext())) {
      // Eat the '.'.
      advance();

      while (isDigit(peek())) {
        advance();
      }
    }

    addToken(NUMBER, Double.parseDouble(source.substring(start, current)));
  }

  private void string() {
    while (!isAtEnd() && peek() != '"') {
      if (peek() == '\n') {
        line++;
      }

      advance();
    }

    if (isAtEnd()) {
      Lox.error(line, "Unterminated string.");
      return;
    }

    // Eat the closing '"'.
    advance();

    String value = source.substring(start + 1, current - 1);
    addToken(STRING, value);
  }

  private boolean isAtEnd() {
    return current >= source.length();
  }

  private boolean isDigit(char c) {
    return c >= '0' && c <= '9';
  }

  private boolean isAlpha(char c) {
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
  }

  private boolean isAlphaNumeric(char c) {
    return isDigit(c) || isAlpha(c);
  }

  private void addToken(TokenType type) {
    addToken(type, null);
  }

  private void addToken(TokenType type, Object literal) {
    String text = source.substring(start, current);
    tokens.add(new Token(type, text, literal, line));
  }

  private char advance() {
    current++;
    return source.charAt(current - 1);
  }

  private boolean match(char expected) {
    if (isAtEnd()) {
      return false;
    }

    if (source.charAt(current) != expected) {
      return false;
    }

    current++;
    return true;
  }

  private char peek() {
    if (isAtEnd()) {
      return '\0';
    }

    return source.charAt(current);
  }

  private char peekNext() {
    if (current + 1 >= source.length()) {
      return '\0';
    }

    return source.charAt(current + 1);
  }

  public List<Token> scanTokens() {
    while (!isAtEnd()) {
      start = current;
      scanToken();
    }

    tokens.add(new Token(EOF, "", null, line));
    return tokens;
  }
}

File: Crafting Interpreters/src/main/scala/Scanner.scala
package com.craftinginterpreters.lox

import collection.mutable.MutableList
import com.craftinginterpreters.lox.TokenType._

object Scanner {
  val keywords = Map(
    "and" -> AND,
    "class" -> CLASS,
    "else" -> ELSE,
    "false" -> FALSE,
    "for" -> FOR,
    "fun" -> FUN,
    "if" -> IF,
    "nil" -> NIL,
    "or" -> OR,
    "print" -> PRINT,
    "return" -> RETURN,
    "super" -> SUPER,
    "this" -> THIS,
    "true" -> TRUE,
    "var" -> VAR,
    "while" -> WHILE
  )
}

class Scanner(source: String) {
  var start = 0
  var current = 0
  var line = 0

  val tokens = new MutableList[Token]()

  def scanTokens(): MutableList[Token] = {
    while (!isAtEnd()) {
      start = current
      scanToken()
    }

    tokens += new Token(EOF, "", None, line)
    tokens
  }

  private def scanToken() = {
    advance() match {
      case '(' => addToken(LEFT_PAREN)
      case ')' => addToken(RIGHT_PAREN)
      case '{' => addToken(RIGHT_BRACE)
      case '}' => addToken(LEFT_BRACE)
      case ',' => addToken(COMMA)
      case '.' => addToken(DOT)
      case '-' => addToken(MINUS)
      case '+' => addToken(PLUS)
      case ';' => addToken(SEMICOLON)
      case '*' => addToken(STAR)
      case '!' => addToken(if (matches('=')) BANG_EQUAL else BANG)
      case '=' => addToken(if (matches('=')) EQUAL_EQUAL else EQUAL)
      case '<' => addToken(if (matches('=')) LESS_EQUAL else LESS)
      case '>' => addToken(if (matches('=')) GREATER_EQUAL else GREATER)
      case ' ' | '\r' | '\t' =>
      case '\n' => line += 1
      case '"' => string()
      case 'o' => if (peek() == 'r') addToken(OR)
      case c if isDigit(c) => number()
      case c if isAlpha(c) => identifier()

      case '/' =>
        if (matches('/')) {
          while (peek() != '\n' && !isAtEnd()) {
            advance()
          }
        } else {
          addToken(SLASH)
        }

      case c =>
        Main.error(line, s"Unterminated character: $c")
    }
  }

  private def identifier() = {
    while (isAlphaNumeric(peek())) advance()

    val value = source.substring(start, current)

    if (Scanner.keywords.contains(value)) {
      addToken(Scanner.keywords(value))
    } else {
      addToken(IDENTIFIER)
    }
  }

  private def number() = {
    while (isDigit(peek())) advance()

    if (peek() == '.' && isDigit(peekNext())) {
      // Eat the '.'
      advance()

      while (isDigit(peek())) advance()
    }

    addToken(NUMBER, Some(source.substring(start, current).toDouble))
  }

  private def string() = {
    while (!isAtEnd() && peek() != '"') {
      if (peek() == '\n')
        line += 1

      advance()
    }

    if (isAtEnd()) {
      Main.error(line, "Unterminated string")
    } else {
      // Eat the closing '"'
      advance()
      addToken(STRING, Some(source.substring(start + 1, current - 1)))
    }
  }

  private def isAtEnd() = {
    current >= source.length
  }

  private def isDigit(c: Char) = {
    c >= '0' && c <= '9'
  }

  private def isAlpha(c: Char) = {
    (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_'
  }

  private def isAlphaNumeric(c: Char) = {
    isDigit(c) || isAlpha(c)
  }

  private def addToken(ttype: TokenType): Unit = {
    addToken(ttype, None)
  }

  private def addToken(ttype: TokenType, literal: Option[Any]): Unit = {
    val text = source.substring(start, current)
    tokens += new Token(ttype, text, literal, line)
  }

  private def advance(): Char = {
    current += 1
    source.charAt(current - 1)
  }

  private def matches(expected: Char): Boolean = {
    if (isAtEnd() || source.charAt(current) != expected)
      return false

    current += 1
    true
  }

  private def peek(): Char = {
    if (isAtEnd())
      '\u0000'
    else
      source.charAt(current)
  }

  private def peekNext(): Char = {
    if (current + 1 >= source.length)
      '\u0000'
    else
      source.charAt(current + 1)
  }
}
